app/forms/admin_reservation_form.rb

```diff
@@ -0,0 +1,105 @@
+class AdminReservationForm
+
+  include ActiveModel::Validations
+  include DateHelper
+
+  REPEAT_OPTIONS = %w(
+    daily
+    weekdays_only
+    weekly
+    monthly
+  ).freeze
+
+  delegate :to_key, :to_model, to: :reservation
+  delegate :category, :reserve_start_date, :reserve_start_hour,
+           :reserve_start_min, :reserve_start_meridian, :duration_mins,
+           :reserve_end_date, :reserve_end_hour, :reserve_end_min,
+           :reserve_end_meridian, :admin_note, :expires?,
+           :expires_mins_before, :reserve_start_at, to: :reservation
+  delegate :assign_times_from_params, to: :reservation
+
+  attr_accessor :repeats, :repeat_frequency, :repeat_end_date
+  attr_reader :reservation
+
+  validates :repeat_frequency, inclusion: { in: REPEAT_OPTIONS, allow_nil: true }
+  validates :repeat_end_date, presence: true, if: :repeats?
+
+  validate :cannot_exceed_max_end_date
+  validate :repeat_end_date_after_initial_date
+
+  def initialize(reservation)
+    @reservation = reservation
+  end
+
+  def assign_attributes(attrs)
+    @reservation.assign_attributes(attrs.except(:repeat_frequency, :repeat_end_date, :repeats))
+    @repeats = attrs[:repeats]
+    @repeat_frequency = attrs[:repeat_frequency]
+    @repeat_end_date = parse_usa_date(attrs[:repeat_end_date])
+  end
+
+  def valid?
+    [@reservation.valid?, super].each {}
+    # errors does not support `merge`
+    @reservation.errors.each do |k, error_messages|
+      errors.add(k, error_messages)
+    end
+    errors.none?
+  end
+
+  def save
+    if valid?
+      # @reservation.save
+      reservations = build_recurring_reservations
+      reservations.map(&:save).all?
```

Comment:
Is there a scenario where one might fail, but not others? Should we wrap this in a transaction and use `save!`?

Reply:
I think it's really unlikely--I can't think of a situation where they wouldn't either succeed or fail all together.